### PR TITLE
Full name for users

### LINKED ISSUE #3055: Full name for users profile
## Feature request
Right now users on rubygems.org only have usernames, so public profiles don't have the ability to show full names to help with SEO index

## Describe the solution you'd like
It would be nice to add full_name to users so it can be used to show on the  public profiles this can help to index profiles for search in the same way that pypi does

## Additional context

Pypi screenshot for reference
image

### PR DESCRIPTION
**Pull request by javier-menendez**
### Description

The goal is to extend user profile data to help SEO visibility searches.

It adds a new attribute: full_name to users, it would be optional.

If not present the user profile remains as it is now otherwise it shows the full name and username

resolves #3055 

### Changes

- New attribute full_name for User model.
- Full name on creation and update.
- Page title for profile show now puts: **Profile of username | RubyGems.org | your community gem host**.
- New @title_for_header_only allows setting title only for page head(browser tab).

### Without full name
image

### With full name
image

### Edit profile
image

### Sign Up
image

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -55,6 +56,9 @@ class User < ApplicationRecord
     unpwn: true,
     allow_nil: true,
     unless: :skip_password_validation?
+
+  validates :full_name, length: { within: 0..20 }, allow_nil: true
```
Comment (simi) [inline app/models/user.rb:60]: Is there any reason to limit name to 20 characters?

File: app/views/profiles/show.html.erb [hunk 2]
```diff
@@ -13,14 +16,24 @@
...
-            class: "t-link--black"
-          )
-        %>
-      </h1>
+      <% if @user.full_name.present? %>
+        <h1 id="profile-name" class="profile__header__name t-display-full-name">
+          <%= link_to(@user.full_name, profile_path(@user.display_id), class: "t-link--black") %>
+        </h1>
+
+        <h1 id="profile-name" class="profile__header__username t-display-username">
```
Comment (simi) [inline app/views/profiles/show.html.erb:24]: :thinking: There are 2x `h1` tags sharing the same `id`. Is that intended?

File: app/views/profiles/show.html.erb [hunk 3]
```diff
@@ -1,3 +1,6 @@
+<% @title = t('.title', username: @user.display_id) %>
+<% @page_title_only = true %>
```
Comment (simi) [inline app/views/profiles/show.html.erb:2]: Isn't there any better way to achieve the same result? Maybe `content_for` could be useful in here.

**Comment by simi** [comment]
Hello! It seems a good idea to introduce optional full name. I have left some initial comments regarding code.

File: app/models/user.rb (hunk 1 above)
Comment (javier-menendez) [reply app/models/user.rb:60]: At first, was because of the space and the current design, that the overflow with spaces takes more than one line, not like the display_id. So that could break the layout like this:

image

I added CSS to get this(can you check it?):

image


So should I put the limit to 40 like the username?

File: app/views/profiles/show.html.erb (hunk 2 above)
Comment (javier-menendez) [reply app/views/profiles/show.html.erb:24]: That wasn't intended, just pushed changes to fix that, thanks :+1:

File: app/views/profiles/show.html.erb (hunk 3 above)
Comment (javier-menendez) [reply app/views/profiles/show.html.erb:2]: There current implementation to change the page title uses this helper:

https://github.com/rubygems/rubygems.org/blob/3b04e8182c773d0d3942e08a5035f5cda623c242/app/helpers/application_helper.rb#L2-L5

https://github.com/rubygems/rubygems.org/blob/3b04e8182c773d0d3942e08a5035f5cda623c242/app/views/layouts/application.html.erb#L4

Makes use of @title to modify the page title, but the problem is that there is no way to only modify page title, it shows the title on the content too here:

https://github.com/rubygems/rubygems.org/blob/3b04e8182c773d0d3942e08a5035f5cda623c242/app/views/layouts/application.html.erb#L107

I introduced @page_title_only so as not to touch other calls. Open to suggestions

Comment (javier-menendez) [reply app/views/profiles/show.html.erb]: Changed to manage that logic in page_title helper, now if we define the instance variable @title_for_header_only the title is only for HTML > head > title, and the @title check on the template is as before

**Comment by simi** [comment]
I'm sorry this felt out of my radar @javier-menendez. Are you still interested in finishing this? It is indeed nice feature to add.

**Comment by javier-menendez** [comment]
> I'm sorry this felt out of my radar @javier-menendez. Are you still interested in finishing this? It is indeed nice feature to add.

@simi yes I am, I'll into the failing check

**Comment by simi** [comment]
> > I'm sorry this felt out of my radar @javier-menendez. Are you still interested in finishing this? It is indeed nice feature to add.
> 
> @simi yes I am, I'll look into the failing check

It seems since you're changing DB you need to update included ERD. Running `bin/rake gen_erd` and committing the new diagram should fix your problem.

**Comment by javier-menendez** [comment]
> > > I'm sorry this felt out of my radar @javier-menendez. Are you still interested in finishing this? It is indeed nice feature to add.
> > 
> > 
> > @simi yes I am, I'll look into the failing check
> 
> It seems since you're changing DB you need to update included ERD. Running `bin/rake gen_erd` and committing the new diagram should fix your problem.

@simi I see thanks :+1: just pushed the changes

File: app/models/user.rb (hunk 1 above)
Comment (simi) [reply app/models/user.rb:62]: I think we should not really limit the name. It would be ok to keep it 255 (default DB varchar size if I remember well).

File: db/schema.rb [hunk 4]
```diff
@@ -126,8 +126,8 @@
     t.bigint "user_id"
     t.text "note"
     t.boolean "status", default: true, null: false
-    t.datetime "created_at", precision: 6, null: false
-    t.datetime "updated_at", precision: 6, null: false
+    t.datetime "created_at", null: false
```
Comment (simi) [inline db/schema.rb:129]: Any idea what's the reason for those unrelated changes?

Comment (javier-menendez) [reply db/schema.rb:129]: Seems related to this https://stackoverflow.com/questions/71411355/rails-7-0-schema-has-arbitrary-precision-6-added-on-my-timestamps

File: app/models/user.rb (hunk 1 above)
Comment (javier-menendez) [reply app/models/user.rb]: I updated the limit :+1:

File: db/schema.rb (hunk 4 above)
Comment (simi) [reply db/schema.rb:129]: Trying to address this at https://github.com/rubygems/rubygems.org/pull/3394.

**Comment by simi** [comment]
:information_source: closed by mistake, can you please rebase now @javier-menendez. Some problems pointed to during review should be fixed already in main branch.

**Comment by javier-menendez** [comment]
@simi I just pushed the merge can you check it? Another thing that we could change is the order in fields forms for new and edit forms WDYT?

**Comment by simi** [comment]
Thanks for the rebase @javier-menendez! This is still on my radar. :eyes:

**Review by simi — APPROVED** [review]

**Comment by simi** [comment]
Thanks for your patience again @javier-menendez. I did some testing locally and it works as intended. What about to make it clear name is optional in forms? Few screenshots for other reviewers.

## new input in sign up form

image


## new input in profile edit form

image

## profile when name is present

image

**Comment by javier-menendez** [comment]
@simi I pushed new changes with the hints can you take a look?

image
image

**Comment by simi** [comment]
Looks good!

**Review by aellispierce — APPROVED** [review]
This looks good to me!

**Review by ericherscovich — COMMENTED** [review]
Looks good to me 😄 🚢

**Comment by simi** [comment]
@javier-menendez I think we're one final rebase (to fix erd.svg) from merge :muscle:.

**Review by indirect — APPROVED** [review]

**Comment by simi** [comment]
@javier-menendez I was able to merge on my own. Thanks for this feature! I'll ping you once it gets deployed.
